test/integration/scheduler_perf/scheduler_perf_test.go

```diff
@@ -467,37 +480,38 @@ func createPods(namespace string, cpo *createPodsOp, clientset clientset.Interfa
 	return podCreator.CreatePods()
 }
 
-// barrierOne blocks until all pods in the given namespace are scheduled.
-func barrierOne(podInformer coreinformers.PodInformer, name string, namespace string, wantCount int) error {
-	for {
+// waitUntilPodsScheduledInNamespace blocks until all pods in the given
+// namespace are scheduled. Times out after 600 seconds because even at the
+// lowest observed QPS of ~10 pods/sec, a 5000-node test should complete.
+func waitUntilPodsScheduledInNamespace(podInformer coreinformers.PodInformer, name string, namespace string, wantCount int) error {
+	return wait.PollImmediate(1*time.Second, 600*time.Second, func() (bool, error) {
 		scheduled, err := getScheduledPods(podInformer, namespace)
 		if err != nil {
-			return err
+			return false, err
 		}
 		if len(scheduled) >= wantCount {
-			break
+			return true, nil
 		}
-		klog.Infof("%s: namespace %s: got %d existing pods, want %d", name, namespace, len(scheduled), wantCount)
-		time.Sleep(1 * time.Second)
-	}
-	return nil
+		klog.Infof("%s: namespace %s: got %d pods, want %d", name, namespace, len(scheduled), wantCount)
+		return false, nil
+	})
 }
 
-// barrier blocks until the all pods in the given namespaces are scheduled.
-func barrier(podInformer coreinformers.PodInformer, name string, namespaces []string, numPodsScheduledPerNamespace map[string]int) error {
+// waitUntilPodsScheduled blocks until the all pods in the given namespaces are
+// scheduled.
+func waitUntilPodsScheduled(podInformer coreinformers.PodInformer, name string, namespaces []string, numPodsScheduledPerNamespace map[string]int) error {
 	// If unspecified, default to all known namespaces.
 	if namespaces == nil {
 		for namespace := range numPodsScheduledPerNamespace {
 			namespaces = append(namespaces, namespace)
 		}
 	}
-
 	for _, namespace := range namespaces {
 		wantCount, ok := numPodsScheduledPerNamespace[namespace]
 		if !ok {
 			return fmt.Errorf("unknown namespace %s", namespace)
 		}
-		if err := barrierOne(podInformer, name, namespace, wantCount); err != nil {
+		if err := waitUntilPodsScheduledInNamespace(podInformer, name, namespace, wantCount); err != nil {
```

Comment:
This is going to have a 600s timeout for each namespace. I think we should have a 600s timeout total. If it's too much refactoring to achieve that, leave it.

Reply:
Added a deadline 1800s deadline per workload that is checked at the start of each op. This way, each workload will take at most 1800s (per workload deadline) + 600s (per op deadline) = 2400s in the worst case, which is a sane limit for some of the large 5000 node tests. LMK what you think.

Reply:
Sounds good, but you should be looking at that context somewhere in this function.

Reply:
Done. Unfortunately, there's no PollImmediateWithContext so we can't cancel `waitUntilPodsScheduledInNamespace`.

Reply:
You could still use the context in the anonymous function and do the exact same thing you are doing in this function.

Reply:
Done.